lib/puppet_pal.rb

```diff
@@ -190,8 +239,31 @@ def self.main(manifest = nil, facts = nil, &block)
 
         rescue => detail
           Puppet.log_exception(detail)
-          raise # exit(1)
+          raise
       end
     end
   end
+
+  def self.assert_non_empty_string(s, what, allow_nil=false)
+    unless s.is_a?(String) && s.length > 0 || (allow_nil && s.nil?)
+      if s.is_a?(String)
+        raise ArgumentError(_("Puppet Pal: %{what} must be a non empty string, got ''") % {what: what})
+      else
+        raise ArgumentError(_("Puppet Pal: %{what} must be a non empty string, got value of type '%{type}'") % {what: what, type: s.class})
+      end
+    end
+  end
+
+  def self.assert_optionally_empty_array(a, what, allow_nil=false)
+    unless a.is_a?(Array) || (allow_nil && a.nil?)
```

Comment:
Perhaps a generic method could be added?
```
def self.assert_type(type, value, what, allow_nil)
  TypeAsserter.assert_instance_of(nil, type, value, allow_nil) { _('Puppet Pal: %{what}') % {what: what} }
end
```
combined with constants such as:
```
T_STRING = PStringType::NON_EMPTY
T_STRING_ARRAY = TypeFactory.array_of(T_STRING)
```
would then give nice readable code like:
```
assert_type(T_STRING, envname, 'envname', true) 
assert_type(T_STRING_ARRAY, modulepath, 'modulepath', true)
```

Reply:
I used this as well